linking error when using `{.compile.}` pragma on .c files with same name but located on different paths

<!-- This is not for feature requests and support related issues. -->

<!--
Any description of the issue that might help others understand
Thanks for taking the time to report it.
-->

### Example

<!--
* include an example, using a code block(s), see below
* provide the necessary reproduction steps: command line flags, operating system, etc
-->

```c
// bar.c
int bar_simple(int a, int b) {
  return a + b;
}
```
```c
// foo/bar.c
int bar_foo(int a, int b) {
  return a * b;
}
```
```nim
{.compile: "bar.c".}
{.compile: "foo/bar.c".}

proc bar_simple(a, b: cint): cint {.importc.}
proc bar_foo(a, b: cint): cint {.importc.}

proc main() =
  echo bar_simple(8, 16)
  echo bar_foo(32, 64)

when isMainModule:
  main()
```

### Actual Output

<!--
* please check whether it's already been fixed in the latest version
* include the actual output using a block, see below 
* if it's not instantly clear, describe what to look for
-->

```
/usr/bin/ld: ~/.cache/nimskull/main_d/bar.c.o: in function `bar_foo':
bar.c:(.text+0x0): multiple definition of `bar_foo'; ~/.cache/nimskull/main_d/bar.c.o:bar.c:(.text+0x0): first defined here
/usr/bin/ld: ~/.cache/nimskull/main_d/@mmain.nim.c.o: in function `main__main_7':
@mmain.nim.c:(.text+0x12b): undefined reference to `bar_simple'
collect2: error: ld returned 1 exit status
```

### Expected Output

<!-- expected output and relevant explanation as to why it should work that way -->

```
24
2048
```

### Possible Solution

Nimskull doesn't concatenate complete path into compiled `.o` files from `{.compile.}` pragma as does with `.nim.c.o` compiled from .nim files
- current behavior: `bar.c` and `foo/bar.c` compile to `bar.c.o` so there is a conflict
- expected behavior: `bar.c` and `foo/bar.c` compile to `@mbar.c.o` and `@mfoo@sbar.c.o` respectively

### Additional Information

compiles on Nim
<!--
* especially if you've done any debugging, or
* found workarounds or almost workarounds

we appreciate help, we're a small bunch of part-time volunteers, so please try to dig in:
* for crashes please attempt to debug: [Debugging the compiler](https://nim-works.github.io/nimskull/intern.html#debugging-the-compiler)
* for suspected regressions: [Bisecting for regressions](https://nim-works.github.io/nimskull/intern.html#bisecting-for-regressions)
-->

--- a/compiler/backend/extccomp.nim
+++ b/compiler/backend/extccomp.nim
@@ -393,6 +393,11 @@ proc toObjFile*(conf: ConfigRef; filename: AbsoluteFile): AbsoluteFile =
   # Object file for compilation
   result = AbsoluteFile(filename.string & "." & CC[conf.cCompiler].objExt)
 
+proc externalObjFile*(conf: ConfigRef; cfile: AbsoluteFile): AbsoluteFile =
+  ## Returns the path of the object file to output for the external C file
+  ## `cfile`.
+  result = toObjFile(conf, completeGeneratedExtFilePath(conf, cfile))
+
 proc addFileToCompile*(conf: ConfigRef; cf: Cfile) =
   conf.toCompile.add(cf)
 

--- a/compiler/front/options.nim
+++ b/compiler/front/options.nim
@@ -1300,6 +1300,19 @@ proc completeGeneratedFilePath*(conf: ConfigRef; f: AbsoluteFile,
   result = subdir / RelativeFile f.string.splitPath.tail
   #echo "completeGeneratedFilePath(", f, ") = ", result
 
+proc completeGeneratedExtFilePath*(conf: ConfigRef, f: AbsoluteFile
+                                  ): AbsoluteFile =
+  ## Returns the absolute file path within the cache directory for file `f`.
+  ## This procedure is meant to be used for external files with names not
+  ## controlled by the compiler -- a sub-directory is used to prevent
+  ## collisions.
+  let subdir = getNimcacheDir(conf.active) / RelativeDir("external")
+  try:
+    createDir(subdir.string)
+  except OSError:
+    conf.quitOrRaise "cannot create directory: " & subdir.string
+  result = subdir / RelativeFile(f.string.splitPath.tail)
+
 proc toRodFile*(conf: ConfigRef; f: AbsoluteFile; ext = RodExt): AbsoluteFile =
   result = changeFileExt(completeGeneratedFilePath(conf,
     withPackageName(conf, f)), ext)

--- a/compiler/sem/pragmas.nim
+++ b/compiler/sem/pragmas.nim
@@ -775,8 +775,10 @@ proc processCompile(c: PContext, n: PNode): PNode =
         found = findFile(c.config, file)
         if found.isEmpty: found = AbsoluteFile file
 
-    let obj = toObjFile(c.config):
-      completeCfilePath(c.config, withPackageName(c.config, found), false)
+    # prepend the package name derived from the file path to `found`, in order
+    # to prevent name collisions when there are multiple external C files with
+    # the same name
+    let obj = externalObjFile(c.config, withPackageName(c.config, found))
     docompile(c, n, found, obj, (if n.len == 2: "" else: result[2].strVal))
   else:
     # too many arguments

--- a/compiler/sem/pragmas.nim
+++ b/compiler/sem/pragmas.nim
@@ -775,7 +775,8 @@ proc processCompile(c: PContext, n: PNode): PNode =
         found = findFile(c.config, file)
         if found.isEmpty: found = AbsoluteFile file
 
-    let obj = toObjFile(c.config, completeCfilePath(c.config, found, false))
+    let obj = toObjFile(c.config):
+      completeCfilePath(c.config, withPackageName(c.config, found), false)
     docompile(c, n, found, obj, (if n.len == 2: "" else: result[2].strVal))
   else:
     # too many arguments

--- a/tests/pragmas/sub/cfunction.c
+++ b/tests/pragmas/sub/cfunction.c
@@ -1,0 +1,3 @@
+int cfunction2(void) {
+  return NUMBER_HERE;
+}

--- a/tests/pragmas/tcompile_pragma.nim
+++ b/tests/pragmas/tcompile_pragma.nim
@@ -1,9 +1,14 @@
 discard """
-  output: '''34'''
+  description: "Tests for the `compile` pragma"
+  targets: "c"
 """
 
 {.compile("cfunction.c", "-DNUMBER_HERE=34").}
+# ensure that compiling a second C file with the same name works:
+{.compile("sub/cfunction.c", "-DNUMBER_HERE=1").}
 
 proc cfunction(): cint {.importc.}
+proc cfunction2(): cint {.importc.}
 
-echo cfunction()
+doAssert cfunction() == 34
+doAssert cfunction2() == 1

